### PR TITLE
Don't allocate more executors to a build than total subjobs

### PR DESCRIPTION
**Pull request by tjlee0909**
This addresses https://github.com/box/ClusterRunner/issues/273

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @tjlee0909 has signed the CLA. Thanks for the pull request!

File: app/master/build_scheduler.py [hunk 1]
```diff
@@ -49,7 +49,13 @@ def needs_more_slaves(self):
...
         :rtype: bool
         """
-        return self._num_executors_allocated < self._max_executors and not self._build._unstarted_subjobs.empty()
+        if self._num_executors_allocated >= self._max_executors:
+            return False
+        if self._build._unstarted_subjobs.empty():
+            return False
+        if self._num_executors_allocated >= len(self._build.all_subjobs()):
+            return False
+        return True
```
Comment (tjlee0909) [inline app/master/build_scheduler.py]: I thought that breaking up the conditionals was much easier to read than all of the 'ands' and 'nots' chained together.

**Comment by tjlee0909** [comment]
I posted a comment about this in https://github.com/box/ClusterRunner/issues/273 , but to reiterate: 

I'm going for low-hanging fruit here with the somewhat common manifestation of this bug where the max_executors is set to much higher than the actual number of subjobs calculated by the atomizer.

File: app/master/build_scheduler.py [hunk 2]
```diff
@@ -49,7 +49,13 @@ def needs_more_slaves(self):
 
         :rtype: bool
         """
-        return self._num_executors_allocated < self._max_executors and not self._build._unstarted_subjobs.empty()
+        if self._num_executors_allocated >= self._max_executors:
+            return False
+        if self._build._unstarted_subjobs.empty():
+            return False
+        if self._num_executors_allocated >= len(self._build.all_subjobs()):
```
Comment (tjlee0909) [inline app/master/build_scheduler.py]: Note to self: maybe should compare to self._build._unstarted_subjobs.qsize() instead.

Comment (josephharrington) [inline app/master/build_scheduler.py]: I think I actually prefer checking the total number of subjobs. What were you thinking would be the advantage of checking `_unstarted_subjobs.qsize()`?

**Comment by josephharrington** [comment]
Hallelujah! :+1:

File: app/master/build_scheduler.py (hunk 2 above)
Comment (tjlee0909) [inline app/master/build_scheduler.py]: In the scenario where the slaves being allocated are trickling in (being released from other builds onto this one), then the length of all subjobs is an overestimate, whereas the unstarted_subjobs.qsize() is more of an exact fit.

It's really not a big deal though. I'm happy with leaving it as is and merging it.

Comment (josephharrington) [inline app/master/build_scheduler.py]: Yeah agreed / not a big deal. :metal:

**Pull request merged**